lib/llvmopencl/UnreachablesToReturns.cc

```diff
@@ -138,106 +144,125 @@ static bool convertUnreachablesToReturns(Function &F) {
-#ifdef DEBUG_CONVERT_UNREACHABLE
-      LLVM_DEBUG(dbgs() << "Unhadled BB Terminator: \n";
-      I->dump();
-#endif
-      assert(0 && "Error: unhandled case in detachBBFromPredecessor\n");
+      LLVM_DEBUG(dbgs() << "Unhandled BB Terminator: \n");
+      LLVM_DEBUG(I->dump());
+      assert(0 && "Error: unexpected BB terminator\n");
     }
   }
 
   for (auto [OldI, NewI] : Replacements) {
     ReplaceInstWithInst(OldI, NewI);
   }
+
+  if (PredSwitch != nullptr) {
+    PredSwitch->removeCase(
+        PredSwitch->findCaseValue(PredSwitch->findCaseDest(&BB)));
+  }
 }
 
 static bool deleteBlocksWithUnreachable(Function &F) {
 
   SmallBBSet UnreachableBBs;
 
-  bool Changed = removeUnreachableSwitchCases(F);
-
   for (Function::iterator I = F.begin(), E = F.end(); I != E; ++I) {
     BasicBlock &BB = *I;
     assert(BB.getTerminator());
     if (isa<UnreachableInst>(BB.getTerminator())) {
-#ifdef DEBUG_CONVERT_UNREACHABLE
-      LLVM_DEBUG(dbgs() << "UNREACHABLE found: deleting BB in "
-                        << F.getName().str() << "\n");
-#endif
+      LLVM_DEBUG(dbgs() << "UNREACHABLE found, deleting BB\n");
+      LLVM_DEBUG(BB.dump());
       UnreachableBBs.insert(&BB);
     }
   }
 
   if (UnreachableBBs.empty())
-    return Changed;
+    return false;
 
-  // check BB predecessors recursively, and disconnect them
-  // from blocks which contain an unreachable
-  SmallBBSet UnreachableBBs2;
-  for (auto *BB : UnreachableBBs) {
-    detachBBFromPredecessor(BB, UnreachableBBs2);
+  // Check BB predecessors recursively, and disconnect them
+  // from blocks which contain an unreachable.
+  SmallBBSet HandledUnreachableBBs;
+  while (!UnreachableBBs.empty()) {
+    BasicBlock *BB = *UnreachableBBs.begin();
+    detachBBFromPredecessor(*BB, UnreachableBBs);
+    UnreachableBBs.erase(BB);
+    HandledUnreachableBBs.insert(BB);
   }
 
-  // all relevant blocks should now be disconnected (have 0 predecessors)
-  // delete them
-  for (auto *BB : UnreachableBBs) {
-    if (BB->hasNPredecessors(0)) {
-#ifdef DEBUG_CONVERT_UNREACHABLE
-      LLVM_DEBUG(dbgs() << "deleting BB: \n");
-      BB->dump();
-#endif
-      BB->eraseFromParent();
-    }
-  }
+  while (!HandledUnreachableBBs.empty()) {
 
-  for (auto *BB : UnreachableBBs2) {
-    if (BB->hasNPredecessors(0)) {
-#ifdef DEBUG_CONVERT_UNREACHABLE
-      LLVM_DEBUG(dbgs() << "deleting BB: \n";
-      BB->dump();
-#endif
-      BB->eraseFromParent();
+    auto CandidateBB = HandledUnreachableBBs.begin();
+
+    // We have to delete the "chains" bottom up to avoid having basic blocks
+    // that refer to the values produced by the predecessors in the stem.
+    while (!isa<UnreachableInst>((*CandidateBB)->getTerminator()))
+      ++CandidateBB;
+    assert(CandidateBB != HandledUnreachableBBs.end());
+    BasicBlock *BB = *CandidateBB;
+
+    LLVM_DEBUG(dbgs() << "Deleting BB: \n");
+    LLVM_DEBUG(BB->dump());
+
+    if (BasicBlock *Pred = BB->getSinglePredecessor()) {
```

Comment:
Is it guaranteed that BBs in the HandledUnreachableBBs don't have multiple predecessors? The `getSinglePredecessor()` return `nullptr` in case of multiple predecessors.

Reply:
Good catch. This was a thinko based on the example I stared.